feat: cluster list fetchers and cluster resource fetcher

tmishina opened the pull request:
- [x] Tick to sign-off your agreement to the [Developer Certificate of Origin (DCO) 1.1](../blob/master/DCO1.1.txt)


## What
<!-- _Provide an overview of the scope of the pull request._ -->

This pull request provides a feature to fetch resources of kubernetes clusters achieves (see #9 in details).
The PR also includes a capability to store lists of kubernetes clusters into an evidence locker from BOM (Bill of Materials) and APIs of cloud service providers.

## Why
<!-- _Provide the business justification for the work included in the pull request._ -->

The resources in a kubernetes cluster contain various types of evidences; for example, `spec` of Pods represents configuration of applications, ConfigMap contains the configuration for the kubernetes cluster itself, and threfore fetching the resources of a kubernetes cluster is important capability for compliance evidence validation of kubernetes clusters.


## How
<!-- _Provide a bulleted list of the changes included in the pull request._ -->

This PR contains two main functionalities.

- cluster list fetcher: store the list of clusters
    - `kube/fetchers/fetch_cluster_list.py`: copy BOM (Bill of Materials) specified in an auditree config file into an evidence locker
    - `ibm_cloud/fetchers/fetch_cluster_list.py`: fetch the list of clusters managed by IBM Cloud ([IBM Kubernetes Service](https://www.ibm.com/cloud/container-service/) or IKS, and [Red Hat Openshift Kubernetes Services](https://www.ibm.com/cloud/openshift) or ROKS) by invoking the `ibmcloud` CLI tool
- cluster resource fetcher: fetch resources from the listed clusters by invoking `kubectl` CLI tool 



## Test
<!-- _Provide a bulleted list of tests included in the pull request and/or tests -->
<!-- performed to validate the work included in the pull request._ -->

- tests of cluster list fetcher (`kubernetes` and `ibm_cloud`)  and cluster resource fetcher were passed for the IBM Cloud clusters (both IKS and ROKS)


## Context
<!-- _Provide a bulleted list of GitHub issues, or any other references (mailing list discussion, etc...) that reviewers can reference -->
<!-- for additional information regarding scope of the pull request._ -->

- #9

tmishina committed:
feat: cluster list fetchers and cluster resource fetcher

alfinkel commented:
@tmishina - I will have a look on Monday.  In the mean time, can you read the DCO and check the box in the PR description if you agree?

tmishina commented:
@alfinkel thank you, I've checked the box of DCO.  and I will perform `git rebase main`.

alfinkel commented on arboretum/common/errors.py:
```diff
@@ -0,0 +1,58 @@
+    @property
+    def stdout(self):
+        """Get standard out text of the command."""
+        return self.__stdout
```
- I'd call this exceptions.py to more closely mirror the naming in [the framework](https://github.com/ComplianceAsCode/auditree-framework/blob/main/compliance/utils/exceptions.py).
- It also seems unnecessary to explicitly use the `@property` decorator here.  
- You should also just use one underscore rather than two for private attributes but I think we don't really need them in this class.

Suggest changing this class to this and renaming as exceptions.py:

```python
# -*- mode:python; coding:utf-8 -*-
# Copyright (c) 2020 IBM Corp. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Common custom exception classes."""


class CommandExecutionError(RuntimeError):
    """System command executed exception class."""

    def __init__(self, cmd, stdout, stderr, returncode):
        """
        Initialize an instance.

        Initialize an instance with the return values of the command.
        """
        self.cmd = cmd
        self.stdout = stdout
        self.stderr = stderr
        self.returncode = returncode
```

alfinkel commented on arboretum/common/utils.py:
```diff
@@ -34,3 +38,25 @@ def parse_seconds(seconds):
+
+
+def run_command(cmd, secrets=None):
+    """Run commands in a system."""
```
```suggestion
    """Execute system command."""
```

alfinkel commented on arboretum/common/utils.py:
```diff
@@ -34,3 +38,25 @@ def parse_seconds(seconds):
+def run_command(cmd, secrets=None):
+    """Run commands in a system."""
+    if type(cmd) == str:
+        cmd = cmd.split(' ')
```
What else would it be if not a string?  Also, you should use `isinstance()` instead of `type()` but I think this entire `if` block is not necessary if `cmd` will always be a string.

tmishina replied:
my intention for `cmd` is that it should accept `str` or list of `str` because `subprocess.Popen` accepts `str` (if `shell=True`) or list of `str` (if `shell=False`).  By accepting both types, user of this function does not need to care about the internal implementation.

To clearly show the intention (and apply your comment about `type` and `isinstance`), I plan to modify the code as follows.

```python
    """
    Execute system command.

    :param cmd: a space-separated string or a list of string
    :param secrets: a text which should be masked in log text.
    :returns: standard output of the command.
    """
    if isinstance(cmd, str):
        cmd = cmd.split(' ')
    elif not isinstance(cmd, list):
        raise TypeError('given command line was neither '
                        f'a space-separated string nor list of string: {cmd}')
```

alfinkel replied:
> my intention for cmd is that it should accept str or list of str because subprocess.Popen accepts str (if shell=True) or list of str (if shell=False)

But that's not what's happening in `run_command` currently.  There's no logic to optionally set the `shell=True` parameter for `Popen` and it will always be acting on a list of arguments anyway based on the previous and proposed versions of `run_command` because  if `cmd` comes in as a string, you turn it into a list of arguments.

1.  If you truly want `run_command` to dynamically either execute `cmd` as a string or `cmd` as a list of arguments then `if isinstance(cmd, str):` is true you should apply `shell=True` as a parameter to the `Popen` call.
2.  However, if you want to continue to break `cmd` into arguments then you should use `shlex.split` rather than `str.split`.

Either solution (1) or (2) will achieve:

> ...By accepting both types, user of this function does not need to care about the internal implementation.

A few other things...

- I think you should look into using [subprocess.run()](https://docs.python.org/3/library/subprocess.html#subprocess.run) rather than directly using `Popen`.
- You can then leverage https://docs.python.org/3/library/subprocess.html#subprocess.CalledProcessError for error handling
- It's not necessary to raise a TypeError if cmd is not a string or list.

tmishina replied:
According to your comments and the document of [subprocess](https://docs.python.org/3.7/library/subprocess.html), I plan to change the implementation as follows.  I really appreciate your comments and suggestions on this, thanks.

- use `subprocess.run()` instead of `subprocess.Popen()`
- assume only `list[str]` for `cmd` to leverage functionality (quoting, etc.) of `subprocess.run()`
- delegate error handling to `subprocess.run()`, and use `subprocess.CalledProcessError` instead of custom exception class `CommandExecutionError`
    - to do so, externalize the masking feature to another function
- other changes (accept further parameters and return stderr in addition to stdout for future use)

```python
def mask_secrets(text, secrets):
    """
    Replace secret words in a text with `***`.

    :param str text: a string which may contain secret words.
    :param list[str] secrets: secret word list.
    :returns: masked text.
    """
    for s in secrets:
        text = text.replace(s, '***')
    return text


def run_command(cmd, input_text=None, timeout=None):
    """
    Execute system command.

    This is a wrapper for `subprocess.run()`.

    Example 1: `run_command(['echo', '-n', 'hello'])` returns `('hello','')`.

    Example 2: `run_command(['cat'], input='hello')` returns `('hello','')`.

    Use `subprocess.run()` if other complicated parameters (e.g., encoding)
    should be specified.

    :param list[str] cmd: command line arguments
    :param str input_text: text for standard input of command
    :param int timeout: timeout for command in seconds
    :raises subprocess.CalledProcessError: if the command finishes with
                                           non-zero returncode.
    :raises subprocess.TimeoutExpires: if timeout expires.
    :raises TypeError: if some of `cmd` element is not a `str`.
    :raises IndexError: if length of `cmd` is zero.
    :returns: a tuple of standard output and standard error of the command.
    """
    cp = subprocess.run(
        cmd,
        input=input_text,
        text=True,
        timeout=timeout,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        check=True,
        shell=False
    )
    return cp.stdout, cp.stderr
```

alfinkel commented on arboretum/common/utils.py:
```diff
@@ -34,3 +38,25 @@ def parse_seconds(seconds):
+    if p.returncode != 0:
+        secrets = secrets or []
+        for s in secrets:
+            cmd = cmd.replace(s, '***')
```
I don't think this is right.  After you split it, `cmd` will be a list which does not have a `replace` function.  Did you test this?  There should be a unit test [here](https://github.com/ComplianceAsCode/auditree-arboretum/blob/main/test/test_common_utils.py) for any code that isn't a fetcher or a check like evidence and common utilities.

alfinkel commented on arboretum/ibm_cloud/README.md:
```diff
@@ -18,10 +18,62 @@ how to include the fetchers and checks from this library in your downstream proj
+
+* Class: [ClusterListFetcher][fetch-cluster-list]
+* Purpose: Write the list of IBM Cloud clusters to the evidence locker.
+* Behavior: Log in to IBM Cloud using `ibmcloud login` command, and save the result of `ibmcloud cs cluster ls` command.
```
This really isn't the behavior content we're looking for.  You shouldn't explicitly state the command that's getting executed because we'll have update this README anytime any sort of change to the command happens.  We may also in the future move away from the command and towards using the API if/when it stabilizes but the behavior of the fetcher would remain the same.  I think something that communicates the behavior in a more broad (generic) way would be better.

tmishina replied:
I've changed the description into more abstract expression.
https://github.com/ComplianceAsCode/auditree-arboretum/commit/89f20fdadcd3382457262f8d93911bcd5bcbd01c#diff-fa9c900ac18eaea9441dab042c0ad0c9

> Behavior: Log in to IBM Cloud and save the list of clusters bound with specified account.

alfinkel commented on arboretum/ibm_cloud/README.md:
```diff
@@ -18,10 +18,62 @@ how to include the fetchers and checks from this library in your downstream proj
+    * List of objects representing the IKS accounts
+    * Each object must have the following values
+      * `account` - a list containing names identifying the IKS account, this will map to an IAM token provided in the credentials file
+* Expected configuration example:
```
We changed the format a little to the READMEs.  ref: https://github.com/ComplianceAsCode/auditree-arboretum/blob/main/arboretum/auditree/README.md.  Update `Expected configuration elements` and `Expected configuration example` and anything else that seems different.

tmishina replied:
applied changes to both [arboretum/ibm_cloud/README.md](https://github.com/ComplianceAsCode/auditree-arboretum/pull/15/files#diff-fa9c900ac18eaea9441dab042c0ad0c9)  and [arboretum/kubernetes/README.md](https://github.com/ComplianceAsCode/auditree-arboretum/pull/15/files#diff-f47957446fca3e1172a44cd6eda6abcc)

alfinkel commented on arboretum/ibm_cloud/README.md:
````diff
@@ -18,10 +18,62 @@ how to include the fetchers and checks from this library in your downstream proj
+     }
+   }
+   ```
+* <a name="expected_credentials"></a>Expected credentials:
````
Same as above, I think we're using "Required credentials:" now and I don't think this needs to be an anchor (yet).

tmishina replied:
removed anchor.

alfinkel commented on arboretum/ibm_cloud/README.md:
```diff
@@ -18,10 +18,62 @@ how to include the fetchers and checks from this library in your downstream proj
+      * `IBM_CLOUD_ACCT_B_API_KEY`
+      * `IBM_CLOUD_ACCT_C_API_KEY`
+
+  * NOTE: [API Keys can be generated using the ibmcloud CLI][ic-api-key-create]. E.g.
```
I think we also need to add a note that stipulates that this fetcher expects that the ibmcloud CLI is installed in order to function.  You'll need to provide a link to ibmcloud docs that describe installing the CLI.  Maybe also mention that in order for this to work when running in CI/CD that consumers of this fetcher will have to ensure that the CLI is installed in the CI/CD execution environment.

alfinkel replied:
Judging by the code, you should also ensure that the Kubernetes service plugin is also installed.  You should make a note of that in the README as well.

tmishina replied:
I mentioned the prerequisites as notes in the [README](https://github.com/ComplianceAsCode/auditree-arboretum/pull/15/files#diff-fa9c900ac18eaea9441dab042c0ad0c9).
Is there better position to describe such a prerequisites?

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_list.py:
```diff
@@ -0,0 +1,70 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""IKS cluster list fetcher."""
```
IBM Cloud is probably better than IKS.  Clearer to folks outside of IBM.

tmishina replied:
thanks, changed.

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_list.py:
```diff
@@ -0,0 +1,70 @@
+        cls.logger = cls.locker.logger.getChild(
+            'ibm_cloud.cluster_list_fetcher'
+        )
+        return cls
```
Seems like you can get rid of this setUpClass altogether and stand up the logger just when you need it below.

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_list.py:
```diff
@@ -0,0 +1,70 @@
+    @store_raw_evidence('ibm_cloud/cluster_list.json')
+    def fetch_cluster_list(self):
+        """Fetch IBM Cloud cluster list."""
+        for account in self.config.get(
+                'org.ibm_cloud.cluster_list.config.account'):
```
```python
        accounts = self.config.get('org.ibm_cloud.cluster_list.config.account')
        for account in accounts:
```

tmishina replied:
applied this, thank you.
(BTW I remove `config` from the key name (`'org.ibm_cloud.cluster_list.config.account'` -> `'org.ibm_cloud.cluster_list.account'`) because it is obviously config without saying config...)

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_list.py:
```diff
@@ -0,0 +1,70 @@
+        """Fetch IBM Cloud cluster list."""
+        for account in self.config.get(
+                'org.ibm_cloud.cluster_list.config.account'):
+            return json.dumps({account: self._get_cluster_list(account)})
```
This doesn't make sense.  You're issuing a return statement inside the for loop so the loop will only execute one time no matter how many accounts you've specified.  Did you test this?

tmishina replied:
I added multi-account test and fixed this issue. Thank you for pointing out this.

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_list.py:
```diff
@@ -0,0 +1,70 @@
+    def _get_cluster_list(self, account):
+
+        # get credential for the account
+        api_key = getattr(self.config.creds['ibm_cloud'], account + '_api_key')
```
use f-strings whenever possible.

```python
        api_key = getattr(self.config.creds['ibm_cloud'], f'{account}_api_key')
```

tmishina replied:
I fixed here and another in [arboretum/kubernetes/fetchers/fetch_cluster_resource.py](https://github.com/ComplianceAsCode/auditree-arboretum/blob/caf5c123a155b2e4e0cfeb3361622e08d9a763fb/arboretum/kubernetes/fetchers/fetch_cluster_resource.py#L110-L112).
https://github.com/ComplianceAsCode/auditree-arboretum/blob/caf5c123a155b2e4e0cfeb3361622e08d9a763fb/arboretum/kubernetes/fetchers/fetch_cluster_resource.py#L110-L112

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_list.py:
```diff
@@ -0,0 +1,70 @@
+        try:
+            cluster_list = run_command(cmd)
+        except CommandExecutionError as e:
+            if e.returncode == 2:  # "2" means no plugin error
```
Are you certain that "2" always means a no plugin error and are you certain that the returncode will be an integer?  BTW this is where you can construct the logger you had in your setUpClass.

alfinkel replied:
If the return code is indeed an integer then your comment of `# "2"...` is a bit confusing.  Perhaps `# RC: 2 == no plugin` might be clearer to future selves.

tmishina replied:
thank you for your suggestion I applied this to two files.

https://github.com/ComplianceAsCode/auditree-arboretum/blob/caf5c123a155b2e4e0cfeb3361622e08d9a763fb/arboretum/ibm_cloud/fetchers/fetch_cluster_list.py#L63-L67

https://github.com/ComplianceAsCode/auditree-arboretum/blob/caf5c123a155b2e4e0cfeb3361622e08d9a763fb/arboretum/kubernetes/fetchers/fetch_cluster_resource.py#L140-L143

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_list.py:
```diff
@@ -0,0 +1,70 @@
+                self.logger.warning(
+                    'Failed to execute "ibmcloud cs" command'
+                    ' - trying to install the cs plugin'
+                )
```
This message is confusing.  Perhaps:

```python
                self.logger.warning(
                    'Kubernetes service plugin missing.  '
                    'Attempting to install plugin...'
                )
```

tmishina replied:
(Please see the reply above)

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_list.py:
```diff
@@ -0,0 +1,70 @@
+                run_command('ibmcloud plugin install kubernetes-service')
+                cluster_list = run_command(cmd)
+            else:
+                raise e
```
just `raise` should suffice.

tmishina replied:
thanks, fixed two lines (same block as https://github.com/ComplianceAsCode/auditree-arboretum/pull/15#discussion_r472798976)

alfinkel reviewed (changes requested):
I reviewed through the "IKS" cluster lists fetcher but you can apply README and other general comments to the entire PR.  Address/answer the questions raised and I'll continue the review of the other fetcher after that.

tmishina committed:
refactor: apply comments on PR#15 on exception class

tmishina committed:
refactor: rename arboretum.common.errors to arboretum.common.exceptions

tmishina committed:
feat: run_command also returns standard error

tmishina committed:
docs: add docs for parameters

tmishina committed:
feat: check arguments with isinstance()

tmishina committed:
fix: bug on replacing secret terms in command line text

tmishina committed:
test: add unit tests for run_command

tmishina committed:
feat: change config format

tmishina committed:
chore: remove setUpClass to simplify code

tmishina committed:
chore: clearify and simplify error handling and its configuration

tmishina committed:
fix: bug on handling multiple accounts

tmishina committed:
chore: use f-string

tmishina committed:
docs: align new README format

tmishina committed:
docs: fix docstring

alfinkel commented:
One other point... I think you should perform a logout (as well as a login) from ibmcloud for each account you process.

tmishina committed:
feat: simplify run_command() and externalize secret masking feature

tmishina committed:
test: for run_command() and mask_secrets()

tmishina committed:
refactor: simplify code

tmishina committed:
docs: documentation update

tmishina committed:
feat: logout from IBM Cloud

tmishina committed:
docs: update documentation to align with current format

tmishina committed:
test: fix test

tmishina commented:
I revised and pushed the branch.  I will perform following operations when the review is completed.

- squash the multiple commits into one commit
- `rebase main` 
- bump version

alfinkel commented:
@tmishina - is this ready for another round of reviews?  Assuming so, but just want to verify.  Planning on reviewing tomorrow (8/21).

tmishina commented:
@alfinkel yes, it is ready. I’d appreciate your comments and suggestions, thanks in advance.

alfinkel commented on arboretum/common/utils.py:
```diff
@@ -34,3 +36,52 @@ def parse_seconds(seconds):
+    Replace secret words in a text with `***`.
+
+    :param str text: a string which may contain secret words.
+    :param list[str] secrets: secret word list.
```
```suggestion
    :param list secrets: secret word list.
```

alfinkel commented on arboretum/common/utils.py:
```diff
@@ -34,3 +36,52 @@ def parse_seconds(seconds):
+        check=True,
+        shell=False
+    )
+    return cp.stdout, cp.stderr
```
With these recent changes in functionality, it doesn't really seem like this function provides much value as a wrapper to `subprocess.run()`.  Suggest removing it and just use `subprocess.run()` explicitly in your fetcher.

alfinkel commented on arboretum/ibm_cloud/README.md:
```diff
@@ -18,10 +18,62 @@ how to include the fetchers and checks from this library in your downstream proj
+### Cluster List
+
+* Class: [ClusterListFetcher][fetch-cluster-list]
+* Purpose: Write the list of IBM Cloud clusters to the evidence locker.
```
```suggestion
* Purpose: Writes a list of IBM Cloud clusters for the specified IBM Cloud accounts to the evidence locker.
```

alfinkel commented on arboretum/ibm_cloud/README.md:
```diff
@@ -18,10 +18,62 @@ how to include the fetchers and checks from this library in your downstream proj
+
+* Class: [ClusterListFetcher][fetch-cluster-list]
+* Purpose: Write the list of IBM Cloud clusters to the evidence locker.
+* Behavior: Log in to IBM Cloud and save the list of clusters bound with specified account.
```
```suggestion
* Behavior: For each IBM Cloud account specified, write the IBM Cloud cluster list to the evidence locker as a single evidence file.  TTL is set to 1 day.
```

- What's the reasoning behind storing all the evidence in one file rather than one file per account?

alfinkel commented on arboretum/ibm_cloud/README.md:
````diff
@@ -18,10 +18,62 @@ how to include the fetchers and checks from this library in your downstream proj
+      }
+    }
+  }
+  ```
````
I think that if we go this route, we may wind up getting repetitive with the IBM Cloud account configurations.  Instead we should just do `org.ibm_cloud.accounts` this way any new fetchers needing the IBM Cloud account config can reuse this.  

Suggest:

* Configuration elements:
   * `org.ibm_cloud.accounts`
     * Required
     * List of accounts (string) 
     * Each account is an arbitrary name describing the IBM Cloud account.  It is used to match to the token 
     provided in the credentials file in order for the fetcher to retrieve content from IBM Cloud for that account.
* Example (required) configuration:

   ```json
   {
     "org": {
       "ibm_cloud": {
         "accounts": ["acct_a", "acct_b"]
       }
     }
   }
   ```

alfinkel commented on arboretum/ibm_cloud/README.md:
````diff
@@ -18,10 +18,62 @@ how to include the fetchers and checks from this library in your downstream proj
+      acct_b_api_key=your-ibm-cloud-api-key-for-acct-b
+      ```
+
+    * NOTE 1: This fetcher requires [IBM Cloud CLI][ibm-cloud-cli] is installed. Its Kubernetes plugin will be automatically installed when the plugin is not installed.
````
Change to:
    
* NOTE (1): In order to function properly this fetcher requires that the [IBM Cloud CLI][ibm-cloud-cli] and the 
    Kubernetes service plugin be installed.  Ensure that something similar to the following is part of your CI/CD execution script:

   ```sh
   echo "Installing IBM Cloud developer tool..."
   curl -sL http://ibm.biz/idt-installer | bash
   echo "Installing the IBM Cloud Kubernetes plugin..."
   ibmcloud plugin install kubernetes-service
   ```

alfinkel commented on arboretum/ibm_cloud/README.md:
````diff
@@ -18,10 +18,62 @@ how to include the fetchers and checks from this library in your downstream proj
+      ```
+
+    * NOTE 1: This fetcher requires [IBM Cloud CLI][ibm-cloud-cli] is installed. Its Kubernetes plugin will be automatically installed when the plugin is not installed.
+    * NOTE 2: [API Keys can be generated using the ibmcloud CLI][ic-api-key-create]. E.g.
````
```suggestion
    * NOTE (2): [API keys can be generated][ic-api-key-create] using the IBM Cloud CLI. Perform something similar to:
```

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_list.py:
```diff
@@ -0,0 +1,77 @@
+"""IBM Cloud cluster list fetcher."""
+
+import json
+import subprocess
```
I think that importing subprocess here underscores my earlier comment about removing the `run_command` wrapper and just running `subprocess.run()` directly.

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_list.py:
```diff
@@ -0,0 +1,77 @@
+class ClusterListFetcher(ComplianceFetcher):
+    """Fetch the list of IBM Cloud clusters."""
+
+    @store_raw_evidence('ibm_cloud/cluster_list.json')
```
Also mentioned earlier but it would be good to know the reasoning behind storing all of the clusters for multiple accounts as one evidence file rather than having a separate cluster list evidence file for each account specified.  The latter approach is more in line with our guidelines.

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_list.py:
```diff
@@ -0,0 +1,77 @@
+    @store_raw_evidence('ibm_cloud/cluster_list.json')
+    def fetch_cluster_list(self):
+        """Fetch IBM Cloud cluster list."""
+        accounts = self.config.get('org.ibm_cloud.cluster_list.account')
```
Adjust based on earlier suggestion in README.

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_list.py:
```diff
@@ -0,0 +1,77 @@
+            run_command(
+                ['ibmcloud', 'login', '--no-region', '--apikey', api_key]
+            )
+        except Exception as e:
```
`Exception` seems a bit broad.  Wouldn't a `subprocess.CalledProcessError` work here?

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_list.py:
```diff
@@ -0,0 +1,77 @@
+        try:
+            run_command(
+                ['ibmcloud', 'login', '--no-region', '--apikey', api_key]
+            )
```
As before replace with `subprocess.run()`.

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_list.py:
```diff
@@ -0,0 +1,77 @@
+                account,
+                mask_secrets(str(e), [api_key])
+            )
+            return None
```
just `return` rather than `return None`.

alfinkel commented on arboretum/ibm_cloud/fetchers/fetch_cluster_list.py:
```diff
@@ -0,0 +1,77 @@
+        try:
+            cluster_list, _ = run_command(cmd)
+        except subprocess.CalledProcessError as e:
+            if e.returncode == 2:  # RC: 2 == no plugin
```
I _think_ I asked this before... How do you know that an RC: 2 means no plugin found?

tmishina replied:
According to my test, `ibmcloud` command returns `2` when a subcommand fails.

```shell
 $ ibmcloud xxxxx
FAILED
'xxxxx' is not a registered command. See 'ibmcloud help'.
 $ echo $?
2
 $ ibmcloud ks xxxxx
FAILED
'xxxxx' is not a registered command. See 'ibmcloud ks help'.

 $ echo $?
2
```

for other errors like login error or network error, `ibmcloud` returns 1.

```shell
 $ ibmcloud ks cluster ls
FAILED
Log in to the IBM Cloud CLI by running 'ibmcloud login'.

 $ echo $?
1
 $ # login, and disconnect from network.
 $ ibmcloud ks cluster ls
FAILED
classic error: Request failed to complete:
Get "https://containers.cloud.ibm.com/global/v1/clusters": dial tcp: lookup containers.cloud.ibm.com on [::1]:53: read udp [::1]:54553->[::1]:53: read: connection refused

Incident ID: 6239f023-fdb0-456b-9d6d-ddba7bb3825f

VPC error: Request failed to complete:
Get "https://containers.cloud.ibm.com/global/v2/vpc/getClusters": dial tcp: lookup containers.cloud.ibm.com on [::1]:53: read udp [::1]:54553->[::1]:53: read: connection refused

Incident ID: 530f4ad3-4501-4b09-9e22-e46e59b9a135

 $ echo $?
1
```

This means that there is possibility to fix error `2` by installing a plugin.  As I wrote in the code, the installation trial will be done at only once; when the command continues to return `2`, the fetcher gave up fixing the issue and report the error message.

As far as I know, there is no document which explicitly indicates that `2` is the subcommand (or plugin) error, and therefore I can change the behaviour to modest one (e.g., just saying "failed; please verify `ibmcloud` and `ibmcloud ks` plugin are installed" and not installing the plugin.)

alfinkel reviewed (changes requested):
@tmishina - Let's focus on the IBM Cloud Cluster List fetcher before moving on to reviewing the Kubernetes fetcher.  I think you have the right pieces, it's just a matter of putting them in the proper places.

- I made a config reorg suggestion that I think will make things better for future fetcher implementations.  Obviously that change propagates to the fetcher code.
- I'd like to settle on whether the proper behavior for this IBM Cloud Cluster List fetcher is to produce one evidence file or multiple evidence files.  As I said in the review, multiple evidence files is more in line with our guidelines but it's not impossible to convince me that one file is necessary here.  I'd just like to know the reasons you chose to do that.
- I really like what you did with installing the Kubernetes service plugin, if one is not already installed.  We hadn't though of that in the past.  I'm wondering if we should extend that behavior to installing the CLI if it is not found as well.  Perhaps we execute a `ibmcloud --version` subprocess and if it returns unfavorably then we perform the CLI install.  What do you think?
- Is the Kubernetes service plugin necessary to get just the cluster list?

Once you address these comments and restructure the config and we can settle on the 1 file vs. multiple files question we can look at ordering the subprocess commands that are being executed in an efficient way.  But like I said, All of the pieces are there.  Nice job.

tmishina commented:
@alfinkel 

Thank you for your detailed review comments!

1. Regarding the `run_command()`, your suggestion is removing it and just use `subprocess.run()` directly.  As you mentioned, `subprocess` functions is no longer encapsulated, and therefore I will remove `run_command()` and change related lines of code not to use it.
2. I agree to your suggestion on reorganizing config items; I will move the account list to just under `ibm_cloud` section.
3. I prefer to use a single evidence file due to simplicity and current usage: the user of cluster resource fetcher will usually specifies same accounts both for the cluster list fetcher and the cluster resource fetcher.  I understand that in future there may be other fetchers or checks to access some portion of accounts.  Do you think that the advantage of multiple evidence file surpasses the simplicity?
4. I added comments on the returncode of `ibmcloud` [here](https://github.com/ComplianceAsCode/auditree-arboretum/pull/15#discussion_r475397597).  Your suggestion about automatically installing `ibmcloud` command would also be convenient for the users, and thus I will implement the functionality.
5. Kubernetes Service is required to all subsubcommands of `ibmcloud ks` including `ibmcloud ks cluster ls`.
6. I will accept your reviews on README.md

tmishina committed:
docs: make description clearer

Co-authored-by: Al Finkelstein <al.finkelstein@ibm.com>

tmishina committed:
feat: remove run_command()

tmishina committed:
docs: further clarify descriptions

tmishina committed:
refactor: not to use common.run_command()

tmishina committed:
feat: change config format

alfinkel commented:
@tmishina - I was chatting with @drsm79 and we've concluded that it would be best to use the [IBM Cloud API](https://cloud.ibm.com/docs?tab=api-docs&) rather than relying on the IBM Cloud CLI and associated plugins.  Doing this removes the overhead of IBM Cloud CLI and plugin installation along with some other benefits.  Sorry for the thrash around the requirements.

tmishina commented:
@alfinkel I investigated steps to fetch resources from an IBM Cloud cluster using its REST API.  Does this make sense?

----


### Get access token and refresh token using API key
https://cloud.ibm.com/apidocs/iam-identity-token-api

```
$ export APIKEY=xxxxx
$ curl -X POST -u 'bx:bx' 'https://iam.cloud.ibm.com/identity/token' -H 'Content-Type: application/x-www-form-urlencoded' -d "grant_type=urn:ibm:params:oauth:grant-type:apikey&apikey=${APIKEY}"
```

returns:

``` json
{
  "access_token": "eyJ...",
  "refresh_token": "OKDMQ...",
  "ims_user_id": 0000000,
  "token_type": "Bearer",
  "expires_in": 3600,
  "expiration": 1598322312,
  "scope": "ibm openid"
}
```

```
$ export TOKEN="Bearer eyJ..."
$ export REFRESH_TOKEN="OKDMQ..."
```

### Get cluster list
https://cloud.ibm.com/apidocs/kubernetes#getclusters

```
$ curl -H "Authorization: Bearer ${TOKEN}" https://containers.cloud.ibm.com/global/v1/clusters
```

returns:

``` json
[
  {
    "id": "bpfn...",
    "serverURL": "https://serverurl:9999",
    "type": "kubernetes",
    ...
  },
  ...
]
```

```
$ export CLUSTER_ID="bpfn...."
$ export MASTER_URL="https://serverurl:9999"
```

### (for IKS clusters only) Get token for cluster
https://cloud.ibm.com/apidocs/kubernetes#getclusterconfig

```
$ curl -v -H "Authorization: ${TOKEN}" -H "X-Auth-Refresh-Token: ${REFRESH_TOKEN}" https://containers.cloud.ibm.com/global/v1/clusters/${CLUSTER_ID}/config > config.zip
```

`config.zip` contains a directory `kubeConfig000000000`, which includes a cert file and a kubeconfig file.

```
$ ls
serverurl.pem   kube-config-servername.yml
```

extracting the bearer token for the cluster.

```
$ export KUBE_TOKEN="Bearer $(cat kubeConfig000000000/kube-config-servername.yml | yq -r '.users[0].user."auth-provider".config."id-token"')"
eyJr...
```
(note: [yq](https://github.com/mikefarah/yq) is a wrapper for yaml files to pass it to `jq` command)


### (for OpenShift clusters only) Get token for cluster
https://cloud.ibm.com/docs/openshift?topic=openshift-access_cluster


```
$ curl "${MASTER_URL}/.well-known/oauth-authorization-server" | jq -r .token_endpoint
https://oauthserver/oauth/token
```

```
$ export OAUTH_SERVER="oauthserver"
```

```
$ curl -vvv -u "apikey:${APIKEY}" -H "X-CSRF-Token: a" "https://${OAUTH_SERVER}/oauth/authorize?client_id=openshift-challenging-client&response_type=token"
*   Trying 169.61.228.138...
...
< Location: https://oauthserver/oauth/token/implicit#access_token=9YoE...&expires_in=86400&scope=user%3Afull&token_type=Bearer
...
```

`access_token=9YoE....` is the token for OpenShift kubernetes server.

```
$ export KUBE_TOKEN="Bearer 9YoE..."
```

### Get resource of a cluster


For example, `Pods` can be retrieved by:

```
curl -k -H "Authorization: ${KUBE_TOKEN}"  "${MASTER_URL}/api/v1/pods"
```

The option `-k` in only required for an OpenShift cluster.

alfinkel commented:
You're going to use Python `requests` rather than curl right?

FYI, the fetcher base class has a [session method](https://github.com/ComplianceAsCode/auditree-framework/blob/main/compliance/fetch.py#L34) that wraps `requests.Session`.  You should use that.

tmishina committed:
feat: utility to get tokens from IAM

tmishina committed:
feat: use IBM Cloud API to get cluster list

tmishina committed:
feat: use IBM Cloud API to get resources of IKS cluster

tmishina commented:
@alfinkel I pushed draft code which uses IBM Cloud API instead of `ibmcloud` command.
In my understanding, I don't need to use the [session method](https://github.com/ComplianceAsCode/auditree-framework/blob/main/compliance/fetch.py#L34) because all of the IBM Cloud APIs support bearer token authentication and are not stateful. Am I correct?

@alfinkel @drsm79 I have two concerns about externalizing common functions regarding IBM Cloud APIs. I created a [utility function file](https://github.com/ComplianceAsCode/auditree-arboretum/pull/15/files#diff-8d800233f90ed427709b4b8c18968a41).
1. Does the position and name of the directory (`arboretum/ibm_cloud/util`) align with the standard of arboretum?
2. `iam.py` is used by both [ibm_cloud category code](https://github.com/ComplianceAsCode/auditree-arboretum/pull/15/files#diff-7969aa7ffd65df2f33a3f39d268c4221) and [kubernetes category code](https://github.com/ComplianceAsCode/auditree-arboretum/pull/15/files#diff-48869a03910283a18790752e17f94de5).  This usage deviates from the architecture described in #9, but from the view point of reducing duplicated code it would be reasonable.  Is it better to write a distinct IAM token retrieval code for `kubernetes` category?

alfinkel commented:
You can put `iam.py` in https://github.com/ComplianceAsCode/auditree-arboretum/tree/main/arboretum/common.

alfinkel commented:
I guess there's still the benefit that you won't need to explicitly set the token for every call as it will be part of the session headers upon construction of the session object.

tmishina commented:
@alfinkel OK, from your comment and [the document of requests](https://requests.readthedocs.io/en/master/user/advanced/) I realized that using `Session` not just eliminates header setting lines of code but also accelerates a series of connections to IBM Cloud API endpoint by its connection pooling feature.  I will update the code to use `Session`.

drsm79 commented:
>     2. `iam.py` is used by both [ibm_cloud category code](https://github.com/ComplianceAsCode/auditree-arboretum/pull/15/files#diff-7969aa7ffd65df2f33a3f39d268c4221) and [kubernetes category code](https://github.com/ComplianceAsCode/auditree-arboretum/pull/15/files#diff-48869a03910283a18790752e17f94de5).  This usage deviates from the architecture described in #9, but from the view point of reducing duplicated code it would be reasonable.  Is it better to write a distinct IAM token retrieval code for `kubernetes` category?

I think we should approach this like the "plugins" for kube, since GCP, AWS, Azure... will have their own iam implementations. So, I disagree with:

> You can put `iam.py` in https://github.com/ComplianceAsCode/auditree-arboretum/tree/main/arboretum/common.

Unless it's named something like `iam_ibm.py` or similar.

tmishina committed:
feat: plug-in for fetching cluster resource

tmishina committed:
feat: use session to get clister list

tmishina committed:
feat: use session to get cluster list

tmishina committed:
feat: use session to get cluster resources (temoral impl)

Currently, ComplianceFetcher.session() does not support renewal of the session,
and therefore raw requests.session() is also used at fetching resources.

tmishina committed:
refactor: move iam.py of ibm_cloud to common directory

tmishina commented:
@alfinkel currently two fetchers use `ComplianceFetcher.session()`.  However, I use raw `requests.session` at [fetching IBM Cloud cluster resource](https://github.com/ComplianceAsCode/auditree-arboretum/pull/15/files#diff-fb24df112adb2d13e1cbd8e877751f41R65) because `ComplianceFetcher.session()` does not support renewal of session. 
(the fetcher connects to two or more servers at once - [IBM Cloud API](https://containers.cloud.ibm.com) and each cluster (`cluster["serverURL"]`))

Is there a way to renew `ComplianceFetcher.session()`? according to the source, it look that session renewal is unavailable.

@drsm79 refactored.
- `arboretum/ibm_cloud/util/iam.py` was moved to `arboretum/common/iam_ibm.py`
- now `fetch_cluster_resource_.py` is pluggable. if a developer want to support a new type of cluster (say `mycsp`), the developer can add a new file `fetch_cluster_resource_mycsp.py` which contains a function `fetch_cluster_resource()`. `fetch_cluster_resource` calls the fucntion by [using importlib](https://github.com/ComplianceAsCode/auditree-arboretum/pull/15/files#diff-48869a03910283a18790752e17f94de5R55-R80).  [fetch_cluster_resource_ibm_cloud.py](https://github.com/ComplianceAsCode/auditree-arboretum/pull/15/files#diff-fb24df112adb2d13e1cbd8e877751f41) can be an example of the plugin for fetch_cluster_resource.

tmishina committed:
refactor: delete unused functions and tests

tmishina committed:
feat: now target resouce should be in plural form

tmishina committed:
feat: support openshift clusters

alfinkel commented:
> ... currently two fetchers use ComplianceFetcher.session(). However, I use raw requests.session at fetching IBM Cloud cluster resource because ComplianceFetcher.session() does not support renewal of session.
(the fetcher connects to two or more servers at once - IBM Cloud API and each cluster (cluster["serverURL"]))

> Is there a way to renew ComplianceFetcher.session()? according to the source, it look that session renewal is unavailable.

I'll add a fix shortly to the ComplianceFetcher.session().

alfinkel commented on arboretum/kubernetes/fetchers/fetch_cluster_resource.py:
```diff
@@ -0,0 +1,133 @@
+                if cltype == 'kubernetes':
+                    resources['kubernetes'] = self._fetch_bom_resource()
+                else:
+                    module_name = (
+                        'arboretum.kubernetes.fetchers.'
+                        f'fetch_cluster_resource_{cltype}'
+                    )
+                    function_name = 'fetch_cluster_resource'
+                    try:
+                        module = import_module(module_name)
+                    except ModuleNotFoundError:
+                        self.logger.error(
+                            'Failed to load plugin for cluster type "%s": %s',
+                            cltype,
+                            module_name
+                        )
+                        raise
+                    try:
+                        fetcher = getattr(module, function_name)
+                    except AttributeError:
+                        self.logger.error(
+                            'Failed to load expected funtion "%s" '
+                            'in module "%s"',
+                            function_name,
+                            module_name
+                        )
+                        raise
+                    resources[cltype] = fetcher(self)
```
I know that I said that I wouldn't look at the cluster resource fetcher until the cluster list was done but what is happening here?  Dynamic loading of modules doesn't seem right.

alfinkel commented on arboretum/kubernetes/fetchers/fetch_cluster_resource_ibm_cloud.py:
```diff
@@ -0,0 +1,159 @@
+"""Feching cluster resouce from IBM Cloud."""
```
..and what's the story here?  This doesn't even include a fetcher class.

alfinkel reviewed (changes requested):
I'm confused as to what you're trying to achieve with the cluster resource "fetcher" work.  Can you explain why you're dynamically loading another module mid fetch and why that module while containing "fetch_" methods doesn't contain a fetch class?

I think we may be at a point where you should split this PR up into two PRs.  One PR for fetching the cluster list and one PR for the cluster resource as it seems like this review is starting to go a bit sideways.

alfinkel commented:
https://github.com/ComplianceAsCode/auditree-framework/pull/57 allows for the fetcher session object to be reset.  Once merged auditree-framework v1.2.6 will have the change.

tmishina commented:
@alfinkel thank you for your suggestion, then shall we split this PR (to be closed) into two PRs?

I have created a new branch `9-cluster-list-fetcher` ([comparison with main](https://github.com/ComplianceAsCode/auditree-arboretum/compare/9-cluster-list-fetcher)) which only contains cluster list fetchers and `iam_ibm.py`.  I will create a new PR for this branch.  In my understanding, we are now on the same page about the cluster list fetchers, and the PR can be merged shortly after some administrative operations (e.g., rebase main, bump version).

When the new PR is merged, I will create another PR for branch `9-cluster-resource-fetcher`, which also contains `fetch_cluster_resource.py`.  Let's continue discussions about the cluster resource fetcher such as [usage of session](https://github.com/ComplianceAsCode/auditree-arboretum/pull/15#issuecomment-681507134) and [plugin capability](https://github.com/ComplianceAsCode/auditree-arboretum/pull/15#pullrequestreview-476834058) in the second new PR.

@drsm79 Please tell me if you have any preference about this change, thanks!

drsm79 commented:
Re. the plugins - wouldn't doing this via different fetchers (which maybe share a base class) suffice? It does mean that people would need to configure multiple fetchers (kube fetcher, openshift fetcher etc..) but I think it's more extensible, and gives us the "plugin" behaviour in a more "native" manner.

alfinkel commented:
> In my understanding, we are now on the same page about the cluster list fetchers, and the PR can be merged shortly after some administrative operations (e.g., rebase main, bump version).

Maybe?  We'll see once you've submitted the new PR whether we're just _some administrative operations_ away from an approval.  BTW, I'm hoping that your new PR will already be rebased and squashed.  As far as the version goes, you don't need to bump it because an auto release is not being performed but you should add an update to the CHANGES.md detailing your changes under an `# UNRELEASED` heading.

tmishina commented:
@alfinkel I created a new PR, which includes the cluster list fetcher part of this PR.
The branch of new PR is rebased to current main branch.  Let's move on to the new PR (#25).
